README
3.5-2 (Nov. 8, 2011): Bug-fix release for R 2.13. Updated qi method to
  correspond with a change to VGAM 0.8-4. This bugfix should not affect Mac
  users.

2.8-3 (May 29, 2007):  Stable release for R 2.4.0-2.5.0.  Fixed bugs in 
  help.zelig(), and summary for multinomial logit, bivariate probit, 
  and bivariate logit with multiple imputation.  First version 
  dependencies are as follows:  
     MASS          7.2-34           
     boot          1.2-27           
     VGAM          0.7-1            
     MCMCpack      0.8-2            
     mvtnorm       0.7-5            
     survival      2.31             
     sandwich      2.0-0            
     zoo           1.2-1            
     coda          0.10-7           
     nnet          7.2-34           
     sna           1.4    

2.8-2 (March 3, 2007):  Stable release for R 2.4.0-2.4.1.  Fixed bug in 
  ARIMA simulation process.  

2.8-1 (February 21, 2007):  Stable release for R 2.4.0-2.4.1.  Made  
  setx() compatible with ordred factor variables (thanks to Mike Ward and 
  Kirill Kalinin).  First order dependencies as in version 2.8-1.  

2.8-0 (February 12, 2007):  Stable release for R 2.4.0-2.4.1.
  Released ARIMA models and network analysis models (least
  squares and logit) for sociomatrices.  First version dependencies
  are as follows:
    MASS          7.2-31           
    boot          1.2-27           
    VGAM          0.7-1            
    MCMCpack      0.7-4            
    mvtnorm       0.7-5            
    survival      2.31             
    sandwich      2.0-0            
    zoo           1.2-1            
    coda          0.10-7           
    nnet          7.2-31           
    sna           1.4    

2.7-5 (December 25, 2006):  Stable release for R 2.4.0-2.4.1.
  Fixed bug related to {\tt names.default()}, summary for multiple
  imputation methods, and prediction for ordinal response models (thanks 
  to Brian Ripley, Chris Lawrence, and Ian Yohai).

2.7-3 (November 9, 2006):  Stable release for R 2.4.0.  Fixed bugs related 
   to R check.

2.7-2 (November 5, 2006):  Stable release for R 2.4.0.  Temporarily 
  removed ARIMA models.
     
2.7-1 (November 3, 2006): Stable release for R 2.4.0. Made
  changes regarding the S4 classes in VGAM. The ARIMA
  model for time series data added by Justin Grimmer.  First level
  dependencies are as follows:
     MASS         7.2-29 
     boot         1.2-26 
     VGAM         0.7-1 
     MCMCpack     0.7-4 
     mvtnorm      0.7-5 
     survival     2.29 
     sandwich     2.0-0 
     zoo          1.2-1 
     coda         0.10-7 

2.6-5 (September 14, 2006): Stable release for R 2.3.0-2.3.1.  Fixed bugs 
  in bivariate logit, bivariate probit, multinomial logit, and 
  model.matrix.multiple (related to version 2.6-4, but not previous 
  versions,  thanks to Chris Lawrence).  First level dependencies are as 
  follows:
     MASS          7.2-27.1         
     boot          1.2-26           
     VGAM          0.6-9            
     MCMCpack      0.7-1            
     mvtnorm       0.7-2            
     survival      2.28               
     sandwich      1.1-1            
     zoo           1.0-6            
     coda          0.10-5 

2.6-4 (September 8, 2006): Stable release for R 2.3.0-2.3.1.  Fixed 
  bugs in {\tt setx()}, and bugs related to {\tt multiple}.  Added 
  instructions for installing Fortran tools for Intel macs.  Added 
  the RxC ecological inference model.

2.6-3 (June 19, 2006): Stable release for R 2.0.0-2.3.1.
  Fixed bug in VDC interface functions, and parse.formula().

2.6-2 (June 7, 2006): Stable release for R 2.0.0-2.3.1. Removed R x C EI.  
  Changed data = list() to data = mi() for multiply-imputed data frames.  
  First level version compatabilities are as for version 2.6-1.

2.6-1 (April 29, 2006): Stable release for R 2.0.0-2.2.1. Fixed major bug 
  in ordinal logit and ordinal probit expected value simulation procedure 
  (does not affect Bayesian ordinal probit). (reported by Ian Yohai) Added 
  the following ecological inference (EI) models: Bayesian hierarchical 
  EI, Bayesian dynamic EI, and RxC EI.  First level version compatabilties 
  (at time of release) are as follows:
     MASS          7.2-24 
     boot          1.2-24     
     VGAM          0.6-8       
     MCMCpack      0.7-1      
     mvtnorm       0.7-2       
     survival      2.24       
     sandwich      1.1-1       
     zoo           1.0-6       
     coda          0.10-5

2.5-4 (March 16, 2006): Stable release for R 2.0.0-2.2.1.  Fixed bug related to 
  windows build.  First-level dependencies are the same as in version 2.5-1.  

2.5-3 (March 9, 2006): Stable release for R 2.0.0-2.2.1.  Fixed bugs related to VDC 
  GUI.  First level dependencies are the same as in version 2.5-1.

2.5-2 (February 3, 2006): Stable release for R 2.0.0-2.2.1.  Fixed bugs related to 
  VDC GUI.  First level dependencies are the same as in version 2.5-1.  

2.5-1 (January 31, 2006): Stable release for R 2.0.0-2.2.1.  Added methods for 
  multiple equation models.  Fixed bugs related to robust estimation and upgrade of 
  sandwich and zoo packages.  Revised setx() to use environments.  Added 
  current.packages() to retrieve version of packages upon which Zelig depends.  
  First level version dependencies are as follows:  
     MASS        7.2-24           
     boot        1.2-24           
     VGAM        0.6-7            
     mvtnorm     0.7-2            
     survival    2.20             
     sandwich    1.1-0            
     zoo         1.0-4            
     MCMCpack    0.6-6            
     coda        0.10-3

2.4-6 (October 27, 2005): Stable release for R 2.0.0-2.2.0.  Fixed bug 
  related to simulation for Bayesian Normal regression.  

2.4-5 (October 18, 2005): Stable release for R 2.0.0-2.2.0.  Updated 
  installation instructions. 

2.4-4 (September 29, 2005): Stable release for R 2.0.0-2.2.0.  Fixed 
  links for help.zelig().

2.4-3 (September 29, 2005): Stable release for R 2.0.0-2.2.0.  

2.4-2 (August 30, 2005): Stable release for R 2.0.0-2.1.1.  Fixed bug in 
  setx() related to as.factor() and I().  Streamlined qi.survreg().

2.4-1 (August 15, 2005): Stable release for R 2.0.0-2.1.1.  Added the 
  following Bayesian models:  factor analysis, mixed factor analysis, 
  ordinal factor analysis, unidimensional item response theory, 
  k-dimensional item response theory, logit, multinomial logit, normal, 
  ordinal probit, Poisson, and tobit.  Also fixed minor bug in
  formula (long variable names coerced to list). 

2.3-2 (August 5, 2005): Stable release for R 2.0.0-2.1.1.  Fixed bug in 
  simulation procedure for lognormal model.

2.3-1 (August 4, 2005): Stable release for R 2.0.0-2.1.1.
  Fixed documentation errors related to model parameterization and code 
  bugs related to first differences and conditional prediction for 
  exponential, lognormal, and Weibull models.  (reported by Alison Post)

2.2-4 (July 30, 2005): Stable release for R 2.0.0-2.1.1.  Revised 
  relogit, additing options for weightig in addition to prior
  correction.  (reported by Martin Plöderl)

2.2-3 (July 24, 2005): Stable release for R 2.0.0-2.1.1.  Fixed bug 
  associated with robust standard errors for negative binomial.

2.2-2 (July 13, 2005): Stable release for R 2.0.0-2.1.1.  Fixed bug in 
  setx().  (reported by Ying Lu)

2.2-1 (July 11, 2005):  Stable release for R 2.0.0-2.1.0.  Revised 
  ordinal probit to use MASS library.  Added robust standard errors 
  for the following regression models: exponential, gamma, logit, 
  lognormal, least squares, negative binomial, normal (Gaussian), 
  poisson, probit, and weibull.

2.1-4 (May 22, 2005):  Stable release for R 1.9.1-2.1.0.  Revised 
  help.zelig() to deal with CRAN build of Windows version.  Added
  recode of slots to lists in NAMESPACE.  Revised install.R script
  to deal with changes to install.packages().  (reported by Dan 
  Powers and Ying Lu)

2.1-3 (May 9, 2005):  Stable release for R 1.9.1-2.1.0.  Revised
  param.lm() function to work with bootstrap simulation.  (reported by 
  Jens Hainmueller)

2.1-2 (April 14, 2005):  Stable release for R 1.9.1-2.1.  Revised 
  summary.zelig().

2.1-1 (April 7, 2005):  Stable release for R 1.9.1-2.1.  Fixed bugs in 
  NAMESPACE and summary.vglm().  

2.0-13 (March 11, 2005): Stable release for R 1.9.1-2.0.1.  Fixed bugs in 
  NAMESPACE and rocplot.Rd. 

2.0-12 (February 20, 2005): Stable release for R 1.9.1-2.0.1.  Added plot 
  = TRUE option to rocplot().  

2.0-11 (January 14, 2005): Stable release for R 1.9.1-2.0.1.  Changed 
  class name for subsettted models from "multiple" to "strata", and 
  modified affected functions.

2.0-10 (January 4, 2005): Stable release for R 1.9.1 and R 2.0.0.  Fixed 
  bug in simulation procedure for ordinal logit.  (Reported by Ian Yohai.)

2.0-9 (October 21, 2004): Stable release for R 1.9.1 and R 2.0.0 (Linux 
  only).  Fixed bugs in NAMESPACE.

2.0-8 (October 18, 2004): Stable release for R 1.9.1 and R 2.0.0
  (Linux only).  Revised Zelig for submission to CRAN.
  
2.0-7 (October 14, 2004): Stable release for R 1.9.1 \emph{and} R
  2.0.0 (Linux only).  Fixed bugs in summary.zelig(), NAMESPACE, and
  assorted bugs related to new R release.  Revised syntax for multiple
  equation models.

2.0-6 (October 4, 2004): Stable release for R 1.9.1. Fixed problem
  with NAMESPACE.

2.0-5 (September 25, 2004): Stable release for R 1.9.1.  Changed
  installation procedure to source install.R from Zelig website.

2.0-4 (September 22, 2004): Stable release for R 1.9.1.  Fixed typo in
  installation directions, implemented NAMESPACE, rationalized
  summary.zelig(), and tweaked documentation for least squares.

2.0-3 (September 1, 2004): Stable release for R 1.9.1.  Fixed bug in
  conditional prediction for survival models.

2.0-2 (August 25, 2004): Stable release for R 1.9.1.  Removed
  predicted values from ls.

2.0-1b (July 16, 2004): Stable release for R 1.9.1.  MD5 checksum
  problem fixed.  Revised plot.zelig() command to be a generic function
  with methods assigned by the model.  Revised entire architecture to
  accept multiply imputed data sets with strata.  Added functions to
  simplify adding models.  Completely restructured reference manual.
  Fixed bugs related to conditional prediction in setx and summarizing
  strata in summary.zelig.

1.1-2 (June 24, 2004): Stable release for R 1.9.1 (MD5
  checksum problem not fixed, but does not seem to cause problems).
  Fixed bug in help.zelig().  (reported by Michael L. Levitan)

1.1-1 (June 14, 2004): Stable release for R 1.9.0.  Revised zelig()
  procedure to use zelig2model() wrappers, revised help.zelig() to use a
  data file with extension .url.tab, and revised setx() procedure to
  take a list of fn to apply to variables, and such that fn = NULL
  returns the entire model.matrix().

1.0-8 (May 27, 2004): Stable release for R 1.9.0.  Fixed bug in
  simulation procedure for survival models.  (reported by Elizabeth
  Stuart)

1.0-7 (May 26, 2004): Stable release for R 1.9.0. Fixed bug in relogit
  simulation procedure.  (reported by Tom Vanwellingham)

1.0-6 (May 11, 2004): Stable release for R 1.9.0.  Fixed bug in
  setx.default, which had previously failed to ignore extraneous
  variables in data frame.  (reported by Steve Purpura)

1.0-5 (May 7, 2004): Replaced relogit procedure with memory-efficient
  version. (reported by Tom Vanwellingham)

1.0-4 (April 19, 2004): Stable release for R 1.9.0.  Added vcov.lm
  method; changed print for summary.relogit.

1.0-2 (April 16, 2004): Testing distribution for R 1.9.0. 

1.0-1 (March, 23, 2004): Stable release for R 1.8.1. 


NEWS.md
> All changes to Zelig are documented here. GitHub issue numbers are given after
each change note when relevant. See <https://github.com/IQSS/Zelig/issues>.
External contributors are referenced with their GitHub usernames when
applicable.

Zelig version 5.1.6
==============================

## Major changes

- All Zelig time series models are deprecated.

## Minor changes

- `predit`, `fitted`, `residuals` now accept arguments. #320

Zelig version 5.1.5
==============================

++++ All Zelig time series models will be deprecated on 1 February 2018 ++++


## Bug fixes

-   Resolved an issue where `odds_ratios` standard errors were not correctly
returned for `logit` and `relogit` models. Thanks to @retrography. #302

-   Zelig 4 compatability wrappers now work for `arima` models. Thanks to
@mbsabath. #280

-   Resolved an error when only `setx` was called with `arima` models Thanks to
@mbsabath. #299

-   Resolved an error when `summary` was called after `sim` for `arima` models.
#305

-   Resolved an error when `sim` is used with differenced first-order
autoregressive models. #307

-   `arima` models return informative error when `data` is not found. #308

## Minor

-   Compatibility with testthat 2.0.0

-   Documentation updated to correctly reflect that `tobit` wraps `AER::tobit`.
#315

-   Package terminology in documentation corrected. #316

Zelig version 5.1-4
==============================

## Major changes

-   Speed improvements made to `relogit`. Thanks to @retrography. #88

-   Returns `relogit` weighted case control method to that described in
King and Langche (2001, eq. 11) and used in the Stata `relogit` implementation.
#295

-   Odds ratios now returned from `summary` with `relogit` models via the
`odds_ratios = TRUE` argument. #302

Zelig version 5.1-3
==============================

## Major changes

-   Roxygen documentation improvements.

## Minor changes and bug fixes

-   `zquantile` with Amelia imputed data now working. #277

-   `vcov` now works with `rq` quantile regression models.

-   More informative error handling for conflicting `timeseries` model
arguments. #283

-   Resolved and issue with `relogit` that produced a warning when the fitted
model object was passed to `predict`. #291


Zelig version 5.1-2
==============================

## Major changes

-   !EXPERIMENTAL! interface function `to_zelig` allows users to convert fitted model
objects fitted outside of Zelig to a Zelig object. The function is called
within the `setx` wrapper if a non-Zelig object is supplied. Currently
only works for models fitted with `lm` and many estimated with `glm` and
`svyglm`. #189

-   `get_se` and `get_pvalue` function wrappers created for `get_se` and
`get_pvalue` methods, respectively. #269

-   If `combine_coef_se` is given a model estimated without multiply imputed
data or bootstraps, an error is no longer returned. Instead a list of the
models' untransformed coefficients, standard errors, and p-values is returned. #268

-   `summary` for `logit` models now accepts the argument `odds_ratios`. When
`TRUE` odds ratio estimates are returned rather than coefficient estimates.
Thanks to Adam Obeng. PR/#270.

- `setx` and `sim` fail informatively when passed ZeligEI objects. #271

## Minor changes and bug fixes

-   Resolved a bug where `weights` were not being passed to `svydesign`
in survey models. #258

-   Due to limited functionality and instability, zelig survey estimations
no return a warning and a link to documentation on how to use `to_survey`
via `setx` to bipass `zelig`. #273

-   Resolved a bug where `from_zelig_model` would not extract fitted model
objects for models estimated using `vglm`. #265

-   `get_pvalue` and `get_se` now work for models estimated using `vglm`. #267

-   Improved `ivreg`, `mlogit`, and getter (#266) documentation.

Zelig version 5.1-1
==============================

## Minor changes

-   Average Treatment Effect on the Treated (ATT) vignette added to the online
documentation <https://docs.zeligproject.org/articles/att.html>

-   Corrected vignette URLs.


Zelig version 5.1-0
==============================

## Major changes

-   Introduce a new model type for instrumental-variable regression: `ivreg`
based on the `ivreg` from the AER package. #223

-   Use the Formula package for formulas. This will enable a common syntax for
multiple equations, though currently in Core Zelig it is only
enhances `ivreg`. #241

-   `zelig` calls now support `update`ing formulas (#244) and `.` syntax for
inserting all variables from `data` on the right-hand side of the formula
#87. See also #247.

-   Arbitrary `log` transformations are now supported in `zelig` calls
(exept for `ivreg` regressors). #225

-   Arbitrary `as.factor` and `factor` transformations are now supported in
`zelig` calls.

-   Restored quantile regression (`model = "rq"`). Currently only supports one
`tau` at a time. #255

-   Added `get_qi` wrapper for `get_qi` method.

-   Added `ATT` wrapper for `ATT` method.

-   `gee` models can now be estimated with multiply imputed data. #263

## Minor changes and bug fixes

-   `zelig` returns an error if `weights` are specified in a model estimated
with multiply imputed data. (not possible before, but uninformative error
returned)

-   Code improvement to `factor_coef_combine` so it does not return a warning
for model types with more than 1 declared class.

-   Reorganize README files to meet new CRAN requirements.

-   Switch `bind_rows` for `rbind_all` in `zquantile` as the latter is depricated.
#255

-   Depends on the survival package in order to enable `setx` for exponential
models without explicitly loading survival. #254

-   `relogit` now only accepts one `tau` per call (similar to `quantile`). Fixed
to address #257.

- Additional unit tests.

Zelig version 5.0-17
==============================

## Major changes

-   New function `combine_coef_se` takes as input a `zelig` model estimated

R/vdc.R
zeligListModels<-function(inZeligOnly=TRUE) {
     if (inZeligOnly) {
    		tmp = ls(envir=asNamespace("Zelig"),pattern="^zelig2")
     } else { 
    		tmp = c( ls(envir=asNamespace("Zelig"),pattern="^zelig2"),
         		apropos("zelig2",mode="function"))
     }
     sub("zelig2","", tmp)
}





zeligInstalledModels<-function(inZeligOnly=TRUE,schemaVersion="1.1") {
  chkpkgs<-function(name)  {
    zd=zeligDescribeModelXML(name,schemaVersion=schemaVersion)
    if (is.null(zd)) {
      return (FALSE)
    }
    zdpd= zeligModelDependency(name)[,1]
    if (is.null(zdpd)) {
      return(TRUE)
    }
    ow=options(warn=-1)
    ret=sapply(zdpd,function(x) require(x,character.only=TRUE)==TRUE)
    options(ow)
    return (ret)
  }
  models<-zeligListModels(inZeligOnly=inZeligOnly)
     # Not being the trusting sort, lets check to see if we can run
     # a dummy formula. If not -- almost always means that something
     # required() is missing
     tmpModels<-sapply(models,chkpkgs)
  models[which(tmpModels)]
}

zeligGetSpecial<-function(modelName) {
	 modelDesc = zeligDescribeModel(modelName)
	 return(modelDesc$parameters[[1]]$specialFunction)
}

zeligModelDependency<-function(modelName,repos="") {
        zd= zeligDescribeModel(modelName)

        if (is.null(zd)) { return (NULL) }

        zdpd=zd[which(names(zd)=="package")]
        
        cbind(sapply(zdpd,function(x)x$name),
                sapply(zdpd,function(x){if (is.null(x$CRAN))
                {rv<-repos} else{rv<-x$CRAN};rv;}))
      }


zeligDescribeModel<-function(name,force=FALSE,schemaVersion="1.1") {
    res=try(eval(call(paste("describe.",name,sep=""))),silent=TRUE)
    if (inherits(res,"try-error")) {
        if (force) {
                res=describe.default()
        } else {
                res=NULL
        }
    }
#    res$name<-name           # only here we have access to the model name, so add it to the list.
    if(!is.null(res)){
   res<-check.full(res,name)
  }
    return(res)
}

zeligDescribeModelXML<-function(modelName,force=FALSE,schemaVersion="1.1") {
	zd = zeligDescribeModel(modelName,force,schemaVersion)
	if (is.null(zd)) {
		return(NULL)
	} else {
		return(zmodel2string(zd))
	}

}


zmodel2string<-function(x) {
     xmlList(x)
}
printZeligSchemaInstance<-function(filename=NULL, serverName=NULL,vdcAbsDirPrefix=NULL){
	# open connection 
	schemaURL<-'http://gking.harvard.edu/zelig';
	if (is.null(serverName)) {
		serverName<-system('hostname -f', intern=TRUE)
	}
	if (is.null(vdcAbsDirPrefix)){
		locationURL<-paste('http://', serverName, '/VDC/Schema/analysis/ZeligInterfaceDefinition.xsd',sep="");
	} else {
		locationURL<-paste('file://', vdcAbsDirPrefix, '/VDC/Schema/analysis/ZeligInterfaceDefinition.xsd',sep="");
	}
	schemaLocation<-paste(schemaURL, ' ', locationURL, sep='');
	con<-"";
	if (!is.null(filename)){
		con<-file(filename,"w");
	}
	cat(file=con, "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<zelig xmlns=\"",schemaURL,"\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xsi:schemaLocation=\"",schemaLocation,"\">\n", sep="");
	mssg<- sapply(zeligInstalledModels(),function(x){cat(file=con,zmodel2string(zeligDescribeModel(x)),sep="")},simplify=FALSE);
	cat(file=con,"\n</zelig>\n",sep="");
}


xmlList<-function(z){
  if(is.null(z))
    return ("")
  mins<-c()
  maxs<-c()
  for(i in 1:length(z$parameters)){
    mins<-c(mins,z$parameters[[i]]$equations[[1]])
    maxs<-c(maxs,z$parameters[[i]]$equations[[2]])
  }
  min<-sum(mins)
  if(any(!is.finite(maxs)))
    max<-Inf
  else
    max<-sum(maxs)
  if(max !=1)
    return("")
  res<-paste("<model name=",'"',z$name,'"'," label=",'"',categories()[[z$category]],'"', sep="")
  if(!(is.na(z$parameters[[1]]$specialFunction)))
    res<-paste(res," specialFunction=",'"',z$parameters[[1]]$specialFunction,'"',sep="")
  res<-paste(res,">\n",sep="")
  res<-paste(res,"<description>",z$description, "</description>\n",sep="")
  url <- paste("http://gking.harvard.edu/zelig/doc/",z$name,".pdf",sep="")
  #if(z$name=="irtkd")
  #  res<-paste(res,"<helpLink url=",'"',"http://gking.harvard.edu/zelig/docs/_TT_irtkd_TT__tex2htm.html",'"',sep="")
  #else
  #res<-paste(res,"<helpLink url=",'"',modelURL(z$name,z$description),'"',sep="")
res<-paste(res,"<helpLink url=",'"',url,'"',sep="")

  res<-paste(res,"/>\n", sep="")
  if(any(!(is.null(z$package)))){
    res<-paste(res,"<packageDependency",sep="")
    if(!(is.na(z$package$name)))
      res<-paste(res," name= ",'"',z$package$name,'"',sep="")
    if(!(is.na(z$package$version)))
      res<-paste(res," version= ",'"',z$package$version,'"',sep="")
    if(!is.null(z$package$CRAN) && !(is.na(z$package$CRAN)))
      res<-paste(res," CRAN= ",'"',z$package$CRAN,'"',sep="")  
    res<-paste(res,"/>\n",sep="")
  }
  res<-paste(res,"<formula minEquations=",'"',min,'"',sep="")
  if(is.finite(max))
    res<-paste(res," maxEquations=",'"',max,'"',sep="")
  if(max==1)
    res<-paste(res," simulEq=",'"',0,'"',sep="")
  res<-paste(res,">\n",sep="")
  
  res<-paste(res,"<equation name=",'"',names(z$parameters)[[1]],'"',">\n",sep="")
  if(!(z$name %in% c("exp","lognorm","weibull"))){    ##we are going to delete this !!!
    if(z$parameters[[1]]$depVar){
      res<-paste(res,"<outcome",sep="")
      if(!is.na(z$parameters[[1]]$specialFunction))                 
        {
          if(is.finite(z$parameters[[1]]$varInSpecialFunction[[2]] ))
            res<-paste(res," maxVar=",'"',z$parameters[[1]]$varInSpecialFunction[[2]],'"',sep="")
          res<-paste(res," minVar=",'"',z$parameters[[1]]$varInSpecialFunction[[1]],'"',sep="")
        }
      else
        {
          if(z$parameters[[1]]$depVar !=TRUE)
            res<-paste(res," minVar=",'"',0,'"'," maxVar=",'"',0,'"',sep="")
        }
      res<-paste(res,">\n")
      for(i in 1:length(modeling.types()$depVar[[z$category]]))
        res<-paste(res,"<modelingType>",modeling.types()$depVar[[z$category]][[i]],"</modelingType>\n",sep="")
      res<-paste(res,"</outcome>\n",sep="")
    }
  } else
  res<-paste(res,durOutput())
                                        #explanatory
  if(z$parameters[[1]]$expVar){
    res<-paste(res,"<explanatory ")
    if(z$parameters[[1]]$expVar == TRUE)
      res<-paste(res," minVar=",'"',1,'"',sep="")
    else
      res<-paste(res," minVar=",'"',0,'"'," maxVar=",'"',0,'"',sep="")
    res<-paste(res,">\n")
    
 
    for(i in 1:length(modeling.types()$expVar[[z$category]]))
      res<-paste(res,"<modelingType>",  modeling.types()$expVar[[z$category]][[i]],"</modelingType>\n",sep="")
    res<-paste(res,"</explanatory>\n",sep="")
  }
  res<-paste(res,"</equation>\n",sep="")
   res<-paste(res,"</formula>\n",sep="")
   if(z$parameters[[1]]$expVar)
     sext<-2
  else
    sext<-0
  res<-paste(res,"<setx maxSetx=",'"',sext,'"',"/>\n",sep="")
  res<-paste(res,"</model>\n")
  return(res)
}



check.full<-function(z,name){
 # we suppose that describe.model pass the check
  z$name<-name
  if(is.null(z$package))
  z$package<-list(name=NA,version=NA, CRAN=NA)
  
  for (i in length(z$parameters)){
  if(is.null(z$parameters[[i]]$specialFunction)) z$parameters[[i]]$specialFunction<-NA
  if(is.null(z$parameters[[i]]$varInSpecialFunction)) z$parameters[[i]]$varInSpecialFunction<-NA
}
 return(z)
  
}


modelURL<-function(modelName,modelDesc){
  baseUrl<-"http://gking.harvard.edu/zelig/docs/"
  spec<-"_TT_"
  res<-paste(baseUrl,spec,modelName,spec,"_",substr(modelDesc,0, 13-nchar(modelName)) ,".html",sep="")
  res<-gsub(".","_",res,fixed=TRUE)
  res<-gsub(" ","_",res,fixed=TRUE)
  res
}

modeling.types <-function(){
  res<-list(
            expVar=list(continuous=c("continuous","discrete","nominal","ordinal","binary"),
              dichotomous=c("continuous","discrete","nominal","ordinal","binary"),
              ordinal=c("continuous","discrete","nominal","ordinal","binary"),
              bounded=c("continuous","discrete","nominal","ordinal","binary"),
              multinomial=c("continuous","discrete","nominal","ordinal","binary"),
              count=c("continuous","discrete","nominal","ordinal","binary"),
              mixed=c("continuous","discrete","nominal","ordinal","binary")),
            depVar=list(
              continuous="continuous",
              dichotomous="binary",
              ordinal="ordinal",
              bounded="continuous",
              multinomial=c("nominal","ordinal"),
              count="discrete",
              mixed=c("continuous","discrete","nominal","ordinal","binary")))
  res
}


## this is a temporary function; Is going to be removed after we change "describe" for this duration models;

durOutput <-function(){
res<-"<outcome minVar=\"1\" maxVar=\"1\" label=\"Duration\">\n<modelingType>continuous</modelingType>\n</outcome>\n<outcome maxVar=\"1\" minVar=\"0\" label=\"Censored\">\n<modelingType>binary</modelingType>\n</outcome>\n"
return (res)
  
}
